### PR TITLE
Detailed validation losses

### PR DESCRIPTION
**Pull request by piercus**

### CODE REVIEW

File: src/lbm/models/lbm/lbm_model.py [hunk 1]
```diff
@@ -93,8 +93,10 @@ def on_fit_start(self, device: torch.device | None = None, *args, **kwargs):
         if self.conditioner is not None:
             self.conditioner.on_fit_start(device=device, *args, **kwargs)
 
-    def forward(self, batch: Dict[str, Any], step=0, batch_idx=0, seed=None, *args, **kwargs):
-
+    def forward_timestep(self, batch: Dict[str, Any], timestep: torch.Tensor, step=0, batch_idx=0, seed=None, *args, **kwargs):
```
Comment (deltheil) [inline src/lbm/models/lbm/lbm_model.py]: Having another forward method feels a bit awkward. Why don't you just introduce an optional timestep in the existing forward method, and just make the switch (if None, default, then sample a timestep)?

File: src/lbm/trainer/trainer.py [hunk 2]
```diff
@@ -175,11 +175,62 @@ def validation_step(self, val_batch: Dict[str, Any], val_idx: int) -> dict:
...
+            selected_timesteps = self.model.selected_timesteps
+            prob = self.model.prob
+
+            out = {}
+
+            # Inspired by what is done in SD3 https://arxiv.org/pdf/2403.03206 
+            #  Section 5.3.3
+            #  "we sample loss levels equidistant in t ∈ (0, 1) and compute validation"
+            #  "loss for each level separately. We then average the loss across all but the last (t = 1) levels"
+            #. NB: In SD3, t = 1 corresponds to timestep 1000 here
```
Comment (deltheil) [inline src/lbm/trainer/trainer.py]:
```suggestion
            # NB: In SD3, t = 1 corresponds to timestep 1000 here
```

File: src/lbm/trainer/trainer.py [hunk 3]
```diff
@@ -175,11 +175,62 @@ def validation_step(self, val_batch: Dict[str, Any], val_idx: int) -> dict:
             seed = seed_from_string("-".join(val_batch[self.model.config.seed_key]))
         else:
             seed = None
-        loss = self.model(val_batch, device=self.device, seed=seed)["loss"]
+        if self.model.timestep_sampling == "custom_timesteps":
+
```
Comment (deltheil) [inline src/lbm/trainer/trainer.py]:
```suggestion
```

File: src/lbm/trainer/trainer.py [hunk 4]
```diff
@@ -175,11 +175,62 @@ def validation_step(self, val_batch: Dict[str, Any], val_idx: int) -> dict:
...
+            out["latent_recon_loss"] = sum(
+                out[f"latent_recon_loss_t{timestep}"] * prob[i]
+                for i, timestep in enumerate(selected_timesteps)
+            )
+
+            out["pixel_recon_loss"] = sum(
+                out[f"pixel_recon_loss_t{timestep}"] * prob[i]
+                for i, timestep in enumerate(selected_timesteps)
+            )
```
Comment (deltheil) [inline src/lbm/trainer/trainer.py:221]: I'm not sure to get the rationale here, at least (correct me if I'm wrong) but this goes beyond the scope of this PR (= logging losses with more details). The way you compute the loss across all timesteps will influence the training, right?

**Review by deltheil — COMMENTED** [review]
Some questions, PTAL

File: src/lbm/trainer/trainer.py (hunk 4 above)
Comment (piercus) [reply src/lbm/trainer/trainer.py:221]: It's only impacting `validation_step` not `training_step`, so it should not impact training, only the validation.

### Before this PR

the val_loss is highly depending on the timestep randomly choosen accross 4 different values (even if it's seeded it's still random). 
The val_loss makes sense, in absolute value, only as an average on a quite-big validation dataset.

### After this PR

The val_loss is more stable and should makes sense at each sample, compared to the average train_loss
It's because we replace the randomly choosen timestep by computing all timesteps each time.

NB: there is still a random part in the way the initial noise image is choosen, but the randomness from the timestep selection is removed.

Comment (deltheil) [reply src/lbm/trainer/trainer.py:221]: > It's only impacting validation_step not training_step, so it should not impact training, only the validation.

True... (I was too much looking at these local changes and got fooled!)

File: src/lbm/models/lbm/lbm_model.py [hunk 5]
```diff
@@ -93,10 +93,7 @@ def on_fit_start(self, device: torch.device | None = None, *args, **kwargs):
         if self.conditioner is not None:
             self.conditioner.on_fit_start(device=device, *args, **kwargs)
 
-    def forward_timestep(self, batch: Dict[str, Any], timestep: torch.Tensor, step=0, batch_idx=0, seed=None, *args, **kwargs):
-        """
-        Forward method for a given timestep
-        """
+    def forward(self, batch: Dict[str, Any], timestep: torch.Tensor | None = None, step=0, batch_idx=0, seed=None, *args, **kwargs):
```
Comment (deltheil) [inline src/lbm/models/lbm/lbm_model.py]: You have been using legacy `Optional` elsewhere, so IMO just avoid using `Tensor | None` here.

Also, how about putting timestep after seed, that you recently introduced? That way there is no/less risk to break some callers?

**Review by deltheil — COMMENTED** [review]
PTAL

**Review by deltheil — APPROVED** [review]
